runtime/lua/vim/diagnostic.lua

```diff
@@ -2591,4 +2591,37 @@ function M.fromqflist(list)
   return diagnostics
 end
 
+--- Returns formatted string with diagnostics for the current buffer.
+--- The severities with 0 diagnostics are left out.
+--- Example `E:2 W:3 I:4 H:5`
+---
+--- To customise appearance, set diagnostic signs text with
+--- ```lua
+--- vim.diagnostic.config({
+---   signs = { text = { [vim.diagnostic.severity.ERROR] = 'Error', ... } }
+--- })
+--- ```
+---
+---@return string
+function M.status()
+  local counts = M.count(0)
```

Comment:
is this really going to work? statusline expressions have access to `g:actual_curbuf` (see help), and could pass it. that avoids weirdness.
```suggestion
function M.status(buf)
  local counts = M.count(buf)
```

Reply:
while evaluating  expression in `%{ }`, the current buffer is set to buffer whose statusline is being drawn, which is what we want for vim.diagnostic.count(0) to work correctly. I think passing in g:actual_curbuf (which is, oddly, a string) is not what we want, we want to draw diagnostics per buffer don't we? from my experiments I found that not passing in g:actual_curbuf is what *I* wanted anyway - otherwise I had the same diagnostic count in each buffer statusline (including help buffers, which does not make sense) due to passing in  g:actual_curbuf.